### ISSUE TITLE
Enable use of SecureDrop server staging VMs as HVMs

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop-workstation#95]
Per comments in #58, the use of [HVMs](https://www.qubes-os.org/doc/hvm/) appears to be the best fit for getting SecureDrop server staging VMs (used for QA) running inside Qubes. This is an epic for tracking progress towards this goal.

Preliminary outline:
- [x] Get Trusty images running as HVMs manually
- [x] Get SecureDrop staging VMs running as HVMs manually (in scope of 5/2-5/16 SecureDrop sprint)
- [x] Document this process (in scope of 6/27-7/11 sprint; see #105)
- [x] Automate build/provisioning process of SecureDrop staging VMs for use inside Qubes

**Comment by eloquence** [comment]
Discussed during standup today: @joshuathayer will attempt an automated build via [Packer](https://www.packer.io/) during this sprint.

**Comment by joshuathayer** [comment]
A bit of an update, here!

After some false starts, I'm finally confident about a scheme to build the staging environment on Qubes. Unfortunately, it's still quite a manual process, though after some iterations it's become easier and easier. It's not clear the the process can ever be fully automated.

@conorsch and I worked last week to come up with schemes to make the SecureDrop developer story tenable even if the staging server build isn't automated.  We started exploring the idea of providing pre-provisioned (or partially-provisioned) staging VMs which developers would download, configure, and run. From time to time, we (FPF) would build and provide new HVM images, probably in sync with SD releases.

Currently we're exploring methods of building those downloadable images. @conorsch and I are working in parallel: Conor is working on building standalone Ubuntu Trusty HVM images, while I'm trying to build the community-supported Trusty template VM. Using HVMs is easier to get going, but lacks the ability to integrate with the rest of the system using native Qubes tools, so running desktop applications in the VM (including terminal emulators) is difficult or impossible, and operations like copy/paste to or from those VMs are impossible. On the other hand, building a proper Trusty VM template is proving difficult, but would allow the staging app and monitor VMs to naturally integrate with the rest of the Qubes system.

All things considered, it might not be a huge deal not to support Qubes tools on the staging VMs: since we can SSH to the running VMs, we can probably interact with them to the extent required (ie, starting and stopping services, watching logs, etc). I'll check in with Conor, but it probably makes sense to continue with the HVM approach, since it's shown to be working already.

**Comment by conorsch** [comment]
@joshuathayer Thanks for the detailed report-back. Yes, the StandaloneVM approach seems sound to me. I've got a branch—which I'll push shortly—that enables use of Standalones for the SD staging environment. There's still a manual setup process for configuring the base HVM that's copied on instance create, so we'll need to collaborate pretty heavily on your docs in order to straighten everything out. At a high level, the process looks like:

1. Download Trusty ISO.
2. Install Trusty on StandaloneVM called `sd-ubuntu-trusty`.
3. Copy the `sd-ubuntu-trusty` twice, as `sd-app-base` and `sd-mon-base`.
4. Boot the `sd-app-base` and `sd-mon-base` VMs, edit their `/etc/hosts` and `/etc/network/interfaces`.
5. Shut down.
6. In SecureDrop repo, run `molecule test -s qubes-staging` and the VMs will be created on the fly. 

There are a few shortcomings, of course. The reboot tasks don't boot the machines again—the developer must run `qvm-start sd-app` in order to bring the machine back up. That's a tolerable shortcoming in my mind, though. There's also a bit of dom0 policy tweaks to be made in order to authorize the sd-dev VM to create other VMs. That's where the Qubes Admin API really shines, though, and I'd very much like to document that clearly so devs become familiar with its capabilities.

**Comment by joshuathayer** [comment]
Awesome, thanks for the report. That doesn't seem like a particularly onerous process- thanks for hammering it out.

Relatedly, I'd like to work on a process for spinning up an SD server *dev* environment on Qubes. I'll make an issue for that.

**Comment by conorsch** [comment]
>  SD server dev environment on Qubes

Extremely optimistic that we _already_ support that setup, given the transition to containers for dev in the SecureDrop repo. As long as the "user" has docker privileges—and docker's installed—it should Just Work™.

**Comment by joshuathayer** [comment]
Hi, a quick update. Conor and I have been working over the past sprint to make the staging build process repeatable and well documented. We're close to having that completed- at this point, we have a documented process (see https://github.com/conorsch/securedrop/blob/qubes-staging-environment/securedrop-staging-qubes.md), which we're refining by repeatedly building from scratch and shaking out the remaining hiccups.

A known issue with the process linked above is that it includes overly-permissive qrexec admin policies for the new `sd-dev` domain: we want to limit it to administrating only securedrop-related VMs, but at the moment it's essentially as capable as`dom0`.

We also still need to address the issue of building and hosting a seucredrop server dev environment on Qubes. Conor and I have talked a bit about it and agree that it's doable, but isn't totally trivial. In particular, we'll have to figure out how most easily to direct traffic to the VM hosting the server: inter-VM networking in Qubes requires a somewhat manual configuration process.

**Comment by eloquence** [comment]
For the 6/27-7/11 sprint we've committed to documenting the setup of staging VMs and, as a stretch goal, documenting the creation of a SecureDrop developer environment.

**Comment by joshuathayer** [comment]
@conorsch and I talked a bit today about the possibility of using a non-dom0 VM for provisioning securedrop-related VMs, which would improve developer and administrator ergonomics. The Qubes admin API facilitates VMs _creation_, but provisioning via salt is less well documented. We found this link which hints that it's possible: https://groups.google.com/forum/#!topic/qubes-users/PtzhBZ8pT4w

I'm leaving the link on this issue since it's the epic for the staging/qubes project, but we may want to consider adding some wiki pages for links like this.

**Comment by eloquence** [comment]
I'm removing further automation of the build process for SD server staging VMs from the near-term backlog for now, since it's not on the critical path for the 0.1.0alpha, as it's mainly a long term convenience for the development environment. If anyone feels strongly otherwise, please speak up.

**Comment by eloquence** [comment]
A few of us are using this environment day-to-day now -- the setup is a bit laborious, of course, and would benefit from further automation. Since this tooling lives in SecureDrop Core, I would motion to close this issue, and to track any further automation improvements as separate issues (ideally with a narrower scope than just "make it magical" ;-).

**Comment by eloquence** [comment]
Closing per discussion on 10/23.
